### DISCUSSION TITLE
What is the OAuth login flow for mobile?

### DISCUSSION THREAD

**Discussion by Probotect0r** [ory/kratos#2434] [Q&A] [answered]
Hello, my team is evaluating Ory Kratos for our application. We have web and mobile apps, and one of our requirements is Single Sign On ("Sign in via Google", etc) support on mobile. I haven't been able to figure out the recommended way to do SSO through Kratos on our mobile applications. Is there any documentation on this?

I am not a mobile developer, so I lack the knowledge in this area. But It seems most apps use a WebView to do SSO. With that in mind, here is what I am thinking the flow will look like: 

1. Start the browser SSO login flow in a WebView in our mobile app by opening `/login?flow=<flowid>` with the appropriate OIDC provider in the request body.
2. Kratos will redirect to the OIDC provider's website so that the user can log in.
3. The OIDC provider will redirect back to Kratos once the user has logged in.
4. Kratos will create a new session based on the info received from the OIDC provider.
5. Kratos redirects to the application front-end (as specified by the redirect_url in the kratos.yml), and additionally sets the `ory_kratos_session` cookie.
6. ???
7. Webview exits and the app (somehow) has access to the session token.

The problem with this is that I don't think apps can access the cookies from the WebView they open. For example, [Apple recommends using a Web Authentication Session](https://developer.apple.com/documentation/authenticationservices/authenticating_a_user_through_a_web_service) for authenticating through a web service, which essentially opens a webview. However, the only thing returned to the app is the URL the webview visited (i.e the redirect to the front-end in the case of Kratos SSO).

So I am wondering how we can go about getting access to the Session token in the app. One idea I had was to have Kratos redirect to an endpoint that will redirect to another endpoint, but it will inspect the session cookie and attach the session token to the URL as a query parameter. Are there any security flaws with doing this? It looks like mobile apps that do OAuth directly use this method since the authorization code returned by the provider is included in the redirect url.

Please share any ideas and suggestions. Thanks.

**Comment by gen1us2k** [comment]
Hello. Ory Kratos has [API flows](https://www.ory.sh/docs/kratos/self-service#api-flows) to support login/registration flows on mobile. In that case, Kratos issues a token for a session, and it can be used against `/sessions/whoami` endpoint to get identity. But I'm not sure about OIDC flows

**Reply by Probotect0r** [reply]
Hey, yes, we know about the API flows for non-browser clients for username/password registration and login. But my question is mainly regarding the OIDC flow on mobile. Can you please share this question with the Ory team? Any guidance would be greatly appreciated. We are evaluating Ory Kratos and the cloud offering but Single Sign On support for mobile clients is important to us.

**Reply by vinckr** [reply]
OIDC-flows for API based clients (such as mobile) are under development here:
https://github.com/ory/kratos/pull/2346
We cant give a concrete timeline when it will be released / done, but feel free to follow & contribute to the linked PR.

**Reply by dmakarenko** [reply]
@vinckr thank you for the suggestion 🙌. As far as I could understand the approach there is to simply allow to exchange an `id_token` to a valid Kratos session token. Which means that the actual OIDC sign-in/sign-up flows are implemented within the native client app.

But in some cases implementing those flows manually could be a bit time consuming especially in comparison with web flows which work almost out of the box. I am aware that some OIDC providers make such integration a breeze by offering an SDK but that means including 3rd party code into your app which may have privacy implications.

Do you think that there might be another alternative (potentially a new feature to build within Kratos) which while being powered mostly by the web flow with redirects running in an in-app browser could support a final redirect to the initiating mobile app by verifying the location of that redirect as well as passing session token as a query param?

**Comment by dmakarenko** [comment]
Thank you @Probotect0r for asking this question. My team has arrived to an exactly same conclusion for a "work-around" of how to fit OIDC-only sign-in/sign-up flows into a native mobile app (iOS): by adding an intermediate redirect to our own location powered under the hood by a trivial service which would capture the `ory_kratos_session` cookie, then verify the final redirect location to make sure it is a deep-link associated with our app by checking its custom URL scheme, and finally redirecting to it with all the cookie details passed as query params so that cookie could be "recreated" in the app and then attached to every authenticated API calls.

Since you've asked about any security flaws, I believe that one potential security flow could be exactly in skipping the verification of the location of the final redirect. Which would allow a malicious actor to leverage that flow from their own website and capture sensitive session token in query params.

P.S. Whenever you have time please let me know if you've decided to go forward with that solution and had any success implementing it. Thank you in advance.

**Reply by vinckr** [reply]
Hey @dmakarenko 
THanks for sharing your workaround! Do you actually use it in production? (I have not tested this or something like it yet)

**Reply by dmakarenko** [reply]
Hi Vincent! We are planning to implement it in a couple of weeks. I'll let you know how it goes upon completion (hopefully successful).

**Reply by Probotect0r** [reply]
I am still concerned about any security flaws with this approach, and since there has been no direction from the Ory team yet, I have decided to wait a bit.

**Comment by vinckr** [comment, marked as answer]
Hello @dmakarenko @Probotect0r  
Ory now supports [OIDC / Social sign-in for native and mobile apps](https://changelog.ory.sh/announcements/social-sign-in-for-native-and-mobile-apps) :tada:
